[Obsolete] Capture frame metrics in fragment monitor

jeremyjiang-dev opened the pull request:
Please review https://github.com/firebase/firebase-android-sdk/pull/3592 instead

jeremyjiang-dev commented:
/test check-changed

visumickey commented:
> The public api surface has changed for the subproject firebase-perf: error: Added class com.google.firebase.perf.metrics.FrameMetricsCalculator [AddedClass]
> 
> Please update the api.txt files for the subprojects being affected by this change by running ./gradlew ${subproject}:generateApiTxtFile. Also perform a major/minor bump accordingly.

Has the public surface changed at all? Can we make sure to have the visibility of APIs rightly set?

visumickey commented on firebase-perf/src/test/java/com/google/firebase/perf/application/FragmentStateMonitorTest.java:
```diff
@@ -84,27 +94,85 @@ public void setUp() {
+
+  @Test
+  public void
+      lifecycleCallbacks_differentFrameMetricsCapturedByFma_logFragmentScreenTraceWithCorrectFrames() {
```
Also, vicecersa where "onResume" method is not called for a fragment and ensure that we do not break/crash when "onPaused" is called.

jeremyjiang-dev replied:
Great ideas! Done!

visumickey commented on firebase-perf/src/test/java/com/google/firebase/perf/metrics/FrameMetricsCalculatorTest.java:
```diff
@@ -0,0 +1,94 @@
+    // Slow frames have duration greater than 16ms and frozen frames have duration greater than
+    // 700ms. The key value pair means (duration, num_of_samples).
+    SparseIntArray sparseIntArray = new SparseIntArray();
+    sparseIntArray.append(5, 3);
```
Can we also add an entry that is neither a slow/frozen frame to ensure that we exclude that as a part of calculation?

jeremyjiang-dev replied:
Sure. Done!

jeremyjiang-dev commented:
> > The public api surface has changed for the subproject firebase-perf: error: Added class com.google.firebase.perf.metrics.FrameMetricsCalculator [AddedClass]
> > Please update the api.txt files for the subprojects being affected by this change by running ./gradlew ${subproject}:generateApiTxtFile. Also perform a major/minor bump accordingly.
> 
> Has the public surface changed at all? Can we make sure to have the visibility of APIs rightly set?

This is a stale comment as I have fixed the issue in a following commit :)

jeremyjiang-dev commented on firebase-perf/src/main/java/com/google/firebase/perf/metrics/FrameMetricsCalculator.java:
```diff
@@ -0,0 +1,87 @@
+ *
+ * @hide
+ */
+public class FrameMetricsCalculator {
```
I cannot add package level visibility to `com.google.firebase.perf.metrics` since `Trace.java` needs to be publicly visible.

jeremyjiang-dev commented on firebase-perf/src/main/java/com/google/firebase/perf/metrics/FrameMetricsCalculator.java:
```diff
@@ -0,0 +1,87 @@
+    int frozenFrames = 0;
+
+    if (arr != null) {
+      SparseIntArray frameTimes = arr[FrameMetricsAggregator.TOTAL_INDEX];
```
Total index: The index in the metrics array where the data for [TOTAL_DURATION](https://developer.android.com/reference/kotlin/androidx/core/app/FrameMetricsAggregator#TOTAL_DURATION()) is stored. We calculate the slow/frozen frames from the TOTAL_DURATION.

The reason that I did not pass the frameMetricsAggregator instance is that we might get the arr from different methods from frameMetricsAggregator, such as getMetrics(), remove(), or reset(). Let me know if you think passing the instance is better.

jeremyjiang-dev commented on firebase-perf/src/test/java/com/google/firebase/perf/application/FragmentStateMonitorTest.java:
```diff
@@ -84,27 +95,85 @@ public void setUp() {
   }
 
+  @Test
+  public void lifecycleCallbacks_sameFrameMetricsCapturedByFma_dropFragmentScreenTrace() {
```
Sg. Changed it to `fragments_whenFrameMetricsIsUnchanged_dropsScreenTrace`.

jeremyjiang-dev commented on firebase-perf/src/main/java/com/google/firebase/perf/application/FragmentStateMonitor.java:
```diff
@@ -73,8 +76,11 @@ public void onFragmentResumed(@NonNull FragmentManager fm, @NonNull Fragment f)
     fragmentToTraceMap.put(f, fragmentTrace);
+
+    FrameMetrics frameMetrics =
+        FrameMetricsCalculator.calculateFrameMetrics(this.frameMetricsAggregator.getMetrics());
```
The `frameMetrics` will not be empty because we are not resetting the `frameMetricsAggregator` before fragment starts. The approach here is to take a snapshot of the frame metrics when a fragment starts and another snapshot  when a fragment finishes. We will calculate the difference between these two snapshots as the fragment's screen performance.

jeremyjiang-dev commented on firebase-perf/src/main/java/com/google/firebase/perf/application/FragmentStateMonitor.java:
```diff
@@ -83,15 +89,36 @@ public void onFragmentPaused(@NonNull FragmentManager fm, @NonNull Fragment f) {
+    }
+    if (frozenFrames > 0) {
+      fragmentTrace.putMetric(Constants.CounterNames.FRAMES_FROZEN.toString(), frozenFrames);
+    }
```
Done!

leotianlizhan replied:
Can you also update `AppStateMonitor` to also use the newer utils like `FrameMetricsCalculator` and `ScreenTraceUtils`?

jeremyjiang-dev replied:
Updated in the new PR https://github.com/firebase/firebase-android-sdk/pull/3592

jeremyjiang-dev commented:
/retest

jeremyjiang-dev commented:
/retest
